rllib/utils/replay_buffers/episode_replay_buffer.py

```diff
@@ -0,0 +1,508 @@
+                batch.
+
+        Returns:
+            The sampled batch (observations, actions, rewards, terminateds, truncateds)
+                of dimensions [B, T, ...].
+        """
+        if num_items is not None:
+            assert batch_size_B is None, (
+                "Cannot call `sample()` with both `num_items` and `batch_size_B` "
+                "provided! Use either one."
+            )
+            batch_size_B = num_items
+        # Use our default values if no sizes/lengths provided.
+        batch_size_B = batch_size_B or self.batch_size_B
+        batch_size_T = batch_size_T or self.batch_size_T
+
+        # Uniformly sample n samples from self._indices (all stores timesteps).
+        index_tuples_idx = []
+        observations = [[] for _ in range(batch_size_B)]
+        actions = [[] for _ in range(batch_size_B)]
+        rewards = [[] for _ in range(batch_size_B)]
+        is_first = [[False] * batch_length_T for _ in range(batch_size_B)]
+        is_last = [[False] * batch_length_T for _ in range(batch_size_B)]
+        is_terminated = [[False] * batch_length_T for _ in range(batch_size_B)]
+        is_truncated = [[False] * batch_length_T for _ in range(batch_size_B)]
+
+        B = 0
+        T = 0
+        idx_cursor = 0
+        while B < batch_size_B:
+            # Ran out of uniform samples -> Sample new set.
+            if len(index_tuples_idx) <= idx_cursor:
+                index_tuples_idx.extend(list(np.random.randint(
+                    0, len(self._indices), size=batch_size_B * 10
+                )))
+
+            index_tuple = self._indices[index_tuples_idx[idx_cursor]]
+            episode_idx, episode_ts = (
+                index_tuple[0] - self._num_episodes_ejected, index_tuple[1]
+            )
+            episode = self.episodes[episode_idx]
+
+            # Starting a new chunk, set continue to False.
+            is_first[B][T] = True
+
+            # Begin of new batch item (row).
+            if len(rewards[B]) == 0:
+                # And we are at the start of an episode: Set reward to 0.0.
+                if episode_ts == 0:
+                    rewards[B].append(0.0)
+                # We are in the middle of an episode: Set reward and h_state to
+                # the previous timestep's values.
+                else:
+                    rewards[B].append(episode.rewards[episode_ts - 1])
+            # We are in the middle of a batch item (row). Concat next episode to this
+            # row from the episode's beginning. In other words, we never concat
+            # a middle of an episode to another truncated one.
+            else:
+                episode_ts = 0
+                rewards[B].append(0.0)
+
+            observations[B].extend(episode.observations[episode_ts:])
+            # Repeat last action to have the same number of actions than observations.
+            actions[B].extend(episode.actions[episode_ts:])
+            actions[B].append(episode.actions[-1])
```

Comment:
The docstring further up says that we append an all-zeros-dummy-action! We should change either.

Reply:
Great catch! Fixed the comment in the class docstr.